### ISSUE TITLE
"Next SFML" Modular Renderer Proposal

### ISSUE THREAD

**Issue by TheMaverickProgrammer** [SFML/SFML#2426]
**Preface**
I had a design that I wanted to share with you guys for a while but wasn't sure if SFML would break away from its API at the time.
But since there's discussion about moving away from "old SFML", I hope you all can keep an open mind.

**Problem**
Support multiple backend renderers for next version of SFML

**Review of Current Prop**
[Link to last proposal.](https://github.com/SFML/SFML/pull/1585)

So one issue I see right away with the prop is using enums for renderers. This means the supported types are hard-coded and new devs cannot BYOB. 
Creating some kind of standard for rendering and passing "extension support" to the hands of 3rd party devs will lower the barrier of entry for contributions towards the many types of renderers SFML wants to support and beyond.
This will also pass the responsibility off the shoulders of the core SFML team so they can focus on new SFML features/bugs.

**Research Similar to Problem**
Please keep in mind that what you see is still very much a prototype but it works very well. And the design choices that follow can be applicable to the goal of achieving renderer modularity for the next SFML.

[Proof of modularity](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/ExampleDemo/Demo.cpp#L24)
In this line, I am creating a variety of renderer options to be able to switch to on the fly.

Let's take a look at a "simple" SFML renderer:
[Pre-configured renderer](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/src/Swoosh/Renderers/SimpleRenderer.h)

We can see there's a draw callback, view change callback, primary output texture target, and let's ignore `onEvent(...)` for now.
Let's see a different renderer:
[Custom renderer](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/ExampleDemo/CustomRenderer.h)

Ignoring the custom structs at the top, we see the same sort of callbacks:
draw, view change, primary output texture target, and these `onEvent(...)` functions.

We have a standard structure of behavior that we know each "renderer" should need. Let's talk a little more about the `onEvent(...)` functions and how it can pertain to the SFML modularity problem. Before we do, let's debrief about the differences in graphics programming and why modularity is tricky.

**Programmable Graphics & Friends**
Since different renderers are just built different ™️, we can expect many different ways to send data to and from the GPU. Vulkan, DirectX, OpenGL, all come to mind. Typically we use a private implementation as we see in the SFML framework. But modern graphics programming have programmable shaders and buffers themselves. This was, from my older understanding, one of the issues SFML had with upgrading to OpenGL ES 2: how do we agree on what the variable names for our vertex and fragment defaults should be? and then for newer OpenGL the same could be said about layouts.

This problem would force programmers to require making their shaders follow a specifically named convention defeating the purpose. And this is only limited to OpenGL. The problem compounds itself when we try to include other shaders like DirectX. 

We need: 
1. The ability to support any rendering backend
2. The ability to let the programmer configure their renderer
3. Also provide a default "working" renderer so that programmers can "just get started"
~~4. Be backwards compatible with SFML 2~~ No longer a constraint 🙂

**Dispatching Arbitrary Events**
Back to my example code, let's talk about the `onEvent(...)` functions. The `onEvent(...)` functions are compile-time enforced by my "self-subscribing dispatcher" design. In order to close the gap between the diamond dependency problem, and allow one base class to be used as the renderer "handle", and allow some magnitude of compile-time checking, I separated the renderer into three parts: 
1. [dispatcher ](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/src/Swoosh/Events/Events.h#L17)
2. [receiver](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/src/Swoosh/Events/Events.h#L54) (trampoline) 
3. [subscriber](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/src/Swoosh/Events/Events.h#L64)

For current-SFML compatibility, I needed to force the renderer to use a base class so anyone using my library can just draw SFML drawables the "old fashioned way". This doesn't need to be necessarily enforced in the modular SFML renderer implementation.

Refer to the renderer implementations at the top of my first post and see that the renderer inherits from a base class that uses types of data to subscribe to. When the IRenderer object submits data, it will trampoline through a type-hashed function pointer back to the specialized `onEvent(...)` function. This hash is created on construction and uses raw function pointers to avoid going onto the heap. Even with some safety checks, I tried to reduce the amount of hops to get to our event handling function.

With this design, we can dispatch arbitrary events we know we are subscribing to. And we don't have to know the renderer implementation when we write our SFML app: [Dispatcher in action](https://github.com/TheMaverickProgrammer/Swoosh/blob/feature/swoosh2/ExampleDemo/Scenes/MainMenuScene.h#L242)

**How to Apply This Knowledge: the process for Next SFML**
1. By comparing modern renderers, we can find a common standard for SFML.
2. With a common standard, we can identify and model common draw operations. Let's call those "draw events".
3. By switching to an event-based renderer, we can pack the event structs with only the data we need to mutate drawing contexts, draw to a target, or even perform operations. 
4. If done similarly to the R&D lib presented, we can supply "pre-configured" simple SFML renderers for various backends.
5. By being de-coupled unlike the enum approach, contributors can easily support new renderers and expand on the draw event data types.
6. Furthermore, we can scaffold a renderer interface that can be a front-facing dispatcher much like this R&D lib that hides the graphics-driver specific details
7. Finally, following step number 6, programmers can then extend the interface and configure their dispatcher's draw steps for granular customization.

**Examples**
To bridge the gap between the ideas here and the prototype linked, here is some very light psuedo-code to show that the design pattern used from the R&D library to illustrate the idea can be adapted to this problem.

_Note: could be any shader format, or ideally SFML adopts cross-platform shader tech_
```glsl
#version 330 core
layout (location = 0) in vec3 aPos;

layout (std140) uniform Matrices
{
    mat4 projection;
    mat4 view;
};
uniform mat4 model;

void main()
{
    gl_Position = projection * view * model * vec4(aPos, 1.0);
}  
```

```cpp
// A custom draw event for our SFML app to use later
struct ModelDataChange : sf::RenderEvent {
    sf::VertexArray data;
    ModelDataChange(const sf::VertexArray& data) : data(data), sf::RenderEvent("ModelDataChanged"){}
};

// Note sf::Renderer inherits sf::IRenderer for programmer's own custom event system.
// Note sf::GCommand can trampoline to the desired Graphics Event System as well (Vulkan impl, OGL impl, etc...)
class MyCustomRenderer : public sf::Renderer<ModelDataChange> {
    bool init{};
    sf::FBO fbo;
    sf::RBO rbo[3];
    unsigned int uboMatrices;
    unsigned int uniformBlockIndex[4];
    sf::VertexArray vaoModel;
    sf::Shaders shaders[4]; // red, blue, green, yellow
public:
    void initRenderPass(sf::GCommand& cmd) override {
        // ASSUME: shaders loaded correctly
        
        for(int i = 0; i < 3; i++) {
            cmd.submit(sf::RenderEvents::ColorAttachment(fbo, rbo[i], i));
        }
        
        for(int i = 0; i < 4; i++) {
            cmd.submit(sf::RenderEvents::GetUniformBlockIndex(shaders[i], "Matrices", &uniformBlockIndex[i]));
            cmd.submit(sf::RenderEvents::UniformBlockBind(shaders[i], &uniformBlockIndex[i], 0));
        }
        
        cmd.submit(sf::RenderEvents::GenBuffer(&uboMatrices, 2*sizeof(sf::Mat4), 0, sf::RenderTypes::StaticDraw));
        // calculate both halves of the matrix, note namespace Glsl is dropped
        sf::Mat4 proj = ...; 
        sf::Mat4 view = ...;
        cmd.submit(sf::RenderEvents::BufferUpdateData(&uobMatrices, 0, proj)); // first half
        cmd.submit(sf::RenderEvents::BufferUpdateData(&uobMatrices, sizeof(sf::Mat4), view)); // second half
        
        cmd.submit(sf::RenderEvents::BindFrameBuffer(fbo));
        init = true;
    }
    
    void execRenderPass(sf::GCommand& cmd) override {
        if(!init) {
            initRenderPass(cmd);
        }
        
        // other draw steps can still be encapsulated inside of classes
        static sf::RenderStates states;
        for(int i = 0; i < 4; i++) {
            sf::Mat4 pos;
            pos.translate( /* place in four corners mathematically */);
            states.shader = shaders[i];
            states.shader->setUniform("model", pos);
            fbo.draw(vaoModel, sf::RenderTypes::Triangles);
        }
    }
    
    // handle any custom draw data our SFML program wants to use
    // this is an overly simplified event data type
    // In production, we would want to include more meaningful data such as
    // the asset by name and keep a hash of all uploaded VAOs...
    void onEvent(const ModelDataChanged& event) override {
        vaoModel = event.vertexArray;
    }
};
```

And the SFML app itself would be very similar:
```cpp
int main() {
    MyCustomRenderer renderer;
    
    sf::VertexArray myVao = LoadModelData1();
    
    // Window now needs at least one default graphic system to support and optionally a custom renderer
    // A builder-pattern would be useful here
    std::optional<sf::RenderWindow> maybeWindow = sf::RenderWindow::Create(sf::VideoMode(200, 200), "SFML")
        .usingGraphicsSupport<sf::Graphics::Vulkan, sf::Graphics::OGL>()
        .usingRenderer(renderer)
        .build();
    
    // Something happened. Ideally, read an error
    if(maybeWindow .has_value()) return -1;
    sf::RenderWindow& window = maybeWindow.value();
    
    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            if (event.type == sf::Event::Closed) {
                window.close();
            }
            else if(event.type == sf::Event::KeyPressed) {
                if(event.key.code == sf::Keyboard::A) { 
                    myVao = LoadModelData1();
                } else {
                    myVao = LoadModelData2();
                }
                
                // Update downstream to our custom renderer
                renderer.submit(ModelDataChanged{myVao});
            }
        }

        window.clear();
        window.display();
    }

    return 0;
}
```

I left out the example implementation for the Graphics Event System but the idea is the same and I included `GCommand` functions in the draw step to give the impression of the idea I'm proposing. e.g. there would need to exist some 
```cpp
using sf::RenderEvents;

// This is just the implementation that comes pre-configured and included with SFML. 
// Programmers can write their own sf::GCommand implementation for the graphics event system.
class VulkanGraphicsSystem : public sf::GCommand<GenUniformBlock, BindUniform, BindAttachment, ...> {
  void onEvent(const GenUniformBlock& event) {
     // handle event at vulkan-level
  }
  ...
}
```
... for Vulkan and so on. 

The template use here might be too much. It's possible we do not need to have the same degree of flexibility as we do at for the programmer's own app, but it's a consideration. We could easily just support every event type as a class function that requires overloading. 

```cpp
// Alt. implementation that doesn't use templates. Requires overloading.
class VulkanGraphicsSystem : public sf::GCommand {
  // required overload
  void onEvent(const GenUniformBlock& event) {
     // handle event at vulkan-level
  }
  
  // required overload
  void onEvent(const BindUniform& event) {
     // handle event at vulkan-level
  }
  ...
}
```

**Conclusion**
The idea is that the details for how to handle various renderers is no longer tightly coupled in SFML itself. This will also allow the SFML team to implement the various OS specializations (windows, inputs, audio) separately from the render pipeline. 

An event-based pipeline can encapsulate common event types shared for all graphics systems and the private implementation can handle each event differently via commands `GCommand`. These private implementations will be easily extendable by using event handling so 3rd parties can contribute easily.

The final step is bridging the legacy API conveniences with the front-facing dispatcher `IRenderer` which provides even further extension for a programmer's own desired rendering pipeline. It can be designed in such a way that common `.draw(...)` utilities still exist in the SFML graphics API but now in a place that can be directed by the programmer as they see fit.

We use these event systems in 2 places: at the graphics system-specific level and optionally at the programmer's app level that has abstracted away the graphics system layer.

The `[ ... ]` denotes that the code is fixed and associated systems are tightly-coupled. Everything else is flexible.

`[OS Specifics -> SFML] -> Graphics Event System -> (optional) App Event System -> [SFML -> Window]`

**Considerations**
**1. Let's get the cons out of the way**

The cons for the programmable graphics systems and the programmer's own app rendering system detailed above is that they use templates.
 
Since the common standard for graphics events should be mostly similar, the pre-made Graphics Event Systems (for Vulkan, OGL, DirectX) for SFML **may be satisfied with traditional classes and virtual functions that simply need to be overload**. This would avoid a lot of unneeded templates and headache. Let's assume that is what we do for the rest of this considerations section of the proposal...

This leaves the programmer's own pipeline to use templates which comes at their own expense with how much custom data they want to dispatch. Most data could be packed into even bigger data structs or switching by enum member variables as tags allowing the programmer's renderer to subscribe to only a few events at most.

And finally, the rewrite. That will have to happen.

**2. Pros**
I feel that this proposal comes with a lot more pros. Some have been listed already before, but I'll list them all here:

- SFML can support various graphics systems
- Contributors can easily support new graphics system with a new `sf::GCommand` implementation
- Programmer's can support new event types at their application-level and handled by the abstraction `sf::IRenderer`
- This also allows for SFML apps to change graphics drivers on-the-fly, supporting the first best one at Window creation time
- This also allows the possibility of also allowing SFML apps to change renderer systems on the fly (see R&D Swoosh2 links)
- Once SFML has this system in place and the common graphics systems are supported, it will no longer be held back by time and can keep up with the newest graphics features easily by updating event data

---
Interested in your thoughts 🙏

**Comment by ChrisThrasher** [comment]
I don't support the notion of user-provided rendering backends. I don't even support the idea multiple 1st party rendering backends. My main reason for this is that it entails a massive increase in the ongoing maintenance burden and complexity of SFML that I don't think justifies the increase in flexibility functionality it entails. Maintaining and testing the existing OpenGL code is hard enough.

Generally it's hard to have interchangeable dependencies. It presents ugly problems like whether the dependencies are picked upon configuration or at runtime. If you pick at configuration then installs of the library must make that decision and users can't change it afterward. If you make that decision at runtime then all dependencies must be present when building which is annoying as well. I've dealt with this on past projects and it was no fun. I hope SFML can avoid this.

**Comment by eXpl0it3r** [comment]
Thank you for writing down your proposal. 🙂 

While we will have to tackle the challenge of coupling, I don't see implementing the renderer(s) as a generic rendering "message bus" as the solution for SFML. It solves the coupling problem, but adds quite a bit of complexity on top, that isn't adequately re-couped.

Dynamic backend selection, is really a nice-to-have feature and shouldn't be a central piece of the puzzle to begin with. As Thrasher mentioned, this doesn't entail just writing C++ code, but you have to do a lot of heavy lifting in dependency management, including dynamic loading of libraries and detection there of. The only point where such a dynamic selection can be useful is, when trying to support a fallback (e.g. Vulkan with OpenGL as fallback), having say DirectX and OpenGL side-by-side is really just a gimmick, as basically no one will go and dynamically switch from one to the other renderer. If we do a full render API evolution, we will probably just have one rendering backend per platform and if something else is supported, you could maybe switch at compile time. I see the appeal, but it's not something crucial to have from the start.

There's little benefit in having dynamic dispatching, when there's just one receiver at the other end, even if the implementieren of the receiver changes, and it can become a much simpler design, if you just have a concrete interface to write code against, instead of some more generic/abstract event structure. Besides that, it's unclear, whether you're really able to cover all the use and edge cases with such an API. What I for example am missing on the current proposal is how to deal with resource management. How are texture loaded (from disk to RAM to VRAM)? Who owns the context? How do you transfer resource between renderers? I quite like event-driven systems and believe it's the right choice for a lot of things in game engines, but even there, I'd say dynamic renderer isn't really as useful.

Rendering has real-time requirements and without proper benchmarks, it's really hard to tell, whether the presented proposal could really live up to the milliseconds scale necessary.

In summary, I like the idea of enabling advanced users to optimize for their own use case and potentially allow for a wider support of renderers through that, however this and dynamic backend selection are sugar coating and topics that can be addressed at a later point. The decoupling of renderer and rendering API can be done a lot simpler, than having event dispatching and also with more known variables on overall support and performance.

As such, I also reject this proposal.

**Comment by ChrisThrasher** [comment]
I appreciate all the thought and effort you put into this, but it's not something we want to move forward with.

**Comment by jsoulier** [comment]
To begin, I completely respect the decision to reject the proposal. It makes sense due to the complexity of supporting something like this. However, with the release of SDL3 and the GPU API, maybe it is worth reconsidering?

A common comparison between libraries is SDL vs SFML (vs GLFW but not relevant). I'm a big fan of the SFML API but as someone who is working with 3D graphics, using SDL makes a bit more sense (for me personally, can't speak for everyone).

I've become fairly familiar with the GPU API in SDL3 and the only complexity I've seen from the user perspective is having to manage shaders yourself (compiling to SPV for Vulkan, DXIL for DirectX, etc). They've been able to keep it fairly agnostic between platforms by supporting a minimal set of features. Maybe it's worth revisiting?
